### DISCUSSION TITLE
Introduce a new device_class for casement windows

### DISCUSSION THREAD

**Discussion by chemelli74** [home-assistant/architecture#525] [Entity Models] [answered]
## Context

In Europe we use casement windows more than up/down windows, more common on US.

Proposal is to create a new device_class for the that variant.
Name is mutuate from material design:

`mdi-window-closed-variant`
`mdi-window-open-variant`

I was also made aware of a discussion about this: [what-the-heck-is-up-with-the-window-icon](https://community.home-assistant.io/t/what-the-heck-is-up-with-the-window-icon/220029/)

## Proposal

PR for new device_class: [#46923](https://github.com/home-assistant/core/pull/46923)

## Consequences

The idea is to allow easy customization for users having a on/off icons that otherwise needs a template for each entity.

**Comment by bdraco** [comment]
HomeKit, alexa, and maybe google assistant would need to be updated to know about the new device class.

I don't think we should let MDI drive naming here. `DEVICE_CLASS_WINDOW_CASEMENT` seems like a better choice.

**Comment by chemelli74** [comment]
> HomeKit, alexa, and maybe google assistant would need to be updated to know about the new device class.

 Maybe it's time to move definitions to a dedicated `device_class.py` and import on all integrations.

> I don't think we should let MDI drive naming here. `DEVICE_CLASS_WINDOW_CASEMENT` seems like a better choice.

Good suggestion, once this is approved, I'll update the PR accordingly. Thx.

Simone

**Reply by chemelli74** [reply]
PR updated ;-)

**Comment by chemelli74** [comment]
Just noticed that discussions should be preferred to issue. 
Should I close this one and move there ?

Simone

**Reply by frenck** [reply]
Migrated 👍

**Comment by iMicknl** [comment]
I don't like the current icon as well, but I wonder what this will do to existing integrations. Will almost all integrations adopt this new device class, or will it only be used as device classes set by users manually?

In essence the device class is the same, but the only difference is the icon where there is a difference in style for US vs EU.

**Reply by chemelli74** [reply]
I think the best approach is to use it as customization class for those that prefer a different icon.

**Comment by dgomes** [comment]
I think this should be handled through localisation just like we should handle units...

**Comment by chemelli74** [comment]
Are we close to a decision ?

Simone

**Comment by frenck** [comment]
While I'm in Europe myself, I'm against this suggestion. The main reason is that we are adding it as a localization feature for an icon... as a device class. That feels wrong to me.

For example, I have a some kind Window sensor from China. Now lets say, I have one, and I ship one to Paulus in the USA. We both pair it with ZHA... what should it show?

The thing is, the device class should not really be a customization feature, for that reason, we have different options (e.g. Lovelace to set an icon).

**Reply by chemelli74** [reply]
> For example, I have a some kind Window sensor from China. Now lets say, I have one, and I ship one to Paulus in the USA. We both pair it with ZHA... what should it show?

The idea is to have a default like it is now, but use `customize.yaml` to select your flavor.

> The thing is, the device class should not really be a customization feature, for that reason, we have different options (e.g. Lovelace to set an icon).

You are right, but if you want different icons based on state (open/closed), there is no way in Lovelace to do so.

Simone

**Reply by frenck** [reply]
Customize.yaml should be an exception to use... we should reduce the need for using that, not increase.

**Reply by chemelli74** [reply]
> Customize.yaml should be an exception to use... we should reduce the need for using that, not increase.

Got the point.

My use is mainly `device_class` customization for door and windows sensors. Plus a couple of `friendly_name` (like "Sun") that are not localized.
If there is a better approach, please share.

Simone

**Reply by chemelli74** [reply]
> For example, I have a some kind Window sensor from China. Now lets say, I have one, and I ship one to Paulus in the USA. We both pair it with ZHA... what should it show?

BTW, isn't the same use case as `door` or `garage_door` device_class ? Same device used in different places allow different icons.

Simone

**Reply by Jc2k** [reply]
Regardless of whether they are comparable situations, a mistake in the past doesn't excuse mistakes in the future.

**Reply by frenck** [reply]
> BTW, isn't the same use case as door or garage_door device_class ?

No, those are distinguishable by the integration. For example, MyQ garage door openers.

**Comment by balloob** [comment, marked as answer]
There is no case for a new device class if there are no integrations that will implement it. Requiring people to opt-in via customize is not acceptable. We don't currently have a way to change the device class of an entity.

**Reply by AndreKR** [reply]
What about this dropdown?

![image](https://github.com/home-assistant/architecture/assets/1188538/980bb94c-5e86-4556-8bd8-1e3effeca123)

You already have to set what the contact actually is, so in the same process you could also set if it's a normal window or a sliding window.

**Reply by frenck** [reply]
@AndreKR Please read the discussion above, as the dropdown you show are device classes... which is what this architectural discussion was about.

This discussion has been closed.

../Frenck

**Reply by AndreKR** [reply]
The way I understood the discussion is that there are two main arguments against this new device class:

1. Device classes are an inherent property of the device. You had an example:

> For example, I have a some kind Window sensor from China. Now lets say, I have one, and I ship one to Paulus in the USA. We both pair it with ZHA... what should it show?

Shipping the device to the USA seems to be a very similar (identical even) case as moving the contact from a window to a door. In both cases I would use the dropdown to change the device class. What's the difference between both cases?

2. Devices have to implement them and since we can't force vendors to implement this new device class, adding it is moot.

But those contact sensors always announce themselves via MQTT discovery as `device_class: door`, so if you use them with a window you *already* have to use the dropdown to set the type of device in HA. What's the difference if it's an existing or a new device class?

**Reply by frenck** [reply]
Hi! @AndreKR,

Maybe I wasn't clear, but this architectural discussion has been concluded in 2021, marked with an answer, closed, and is no longer in active discussion.

../Frenck

**Reply by AndreKR** [reply]
I'm not advocating to introduce a new device class, I trust that the people who know more about HA's architecture have good reasons to not do that.

I'm just asking what those reasons were (just to understand the issue) and if there is an alternative approach to deal with the window icon in particular and with devices that are used in multiple different positions (like door/window/cabinet/etc.) in general.

**Comment by caco3** [comment]
Sorry to pop up this issue again...

I also live in Europe and am irritated about the US window visualization.
Do I understand it right that the dropdown list is directly coupled to the device class? Does this really make sense?
There is already a nice list of representations (see https://www.home-assistant.io/integrations/cover/#device-class). Why can't it be extended with the window-varant style?

Alternatively, how can one customize it? I couldn't figure it out yet! Simply changing the icon does not help since it then no longer is dynamically adjusted based on the state. And having to create an extra template button/cover per window seems rater complicated for normal users (especially since there is the real cool dropdown list).


Side note: Beside of the casement windows, it also would be nice to have roof window icons, eg.
![grafik](https://github.com/user-attachments/assets/7b21d42a-e23c-4916-8e76-290c75bf8b11) resp. ![grafik](https://github.com/user-attachments/assets/0bd05746-0ce1-4ed4-ac98-0769a3aa7613)

But that's also missing in the icons list.

**Reply by abmantis** [reply]
Hey @caco3 !

While your request is valid from a feature request PoV (personally, I would also like to be able to define a different icon set for sensors), this proposal here is not the proper approach to achieve that as previously stated when it was closed.

If want to implement that feature and have a proposal on how to do it, please open a new architecture discussion for it. Otherwise, if you want to make a feature request instead, I would ask you to use the [feature request](https://community.home-assistant.io/c/feature-requests) section of our forum instead.

Thanks!

**Reply by caco3** [reply]
Thanks for your response!

Do you have an answer to my question how to do it as of now?

There is already a 4 years old FR: https://community.home-assistant.io/t/custom-device-classes/368238, no point to open another one.

And for the reference: There is already an icon request for the roof window: https://github.com/Templarian/MaterialDesign/issues/7464

**Reply by abmantis** [reply]
It isn't possible AFAIK to do it directly, for now.

**Reply by mundschenk-at** [reply]
It can be done via https://github.com/Mariusthvdb/custom-ui with something like this in `customize_glob.yaml`:

```yaml
binary_sensor.*window*is_open:
    templates:
        icon: >
            if (state === 'on') return 'mdi:window-open-variant';
            return 'mdi:window-closed-variant';
```
